modules/packages/LinearAlgebra.chpl

```diff
@@ -2490,6 +2490,270 @@ proc cosm(A: []) throws {
+  }
+
+  var n = D.dim(0);
+  var est;
+  if (t >= n) {
+    est = norm(A);
+  }
+  else {
+    est = _onenormest(A);
+  }
+
+  return est;
+}
+
+proc _onenormest(A: [?D], param t=2,param itmax=5) throws {
+  if (itmax < 2) {
+    throw new LinearAlgebraError("_onenormest(): At least two iterations are required");
+  }
+  if (t < 1) {
+    throw new LinearAlgebraError("_onenormest(): At least one column is required");
+  }
+
+  var n = D.dim(0);
+
+  if (t <= n) {
+    throw new LinearAlgebraError("_onenormest(): t can't be larger than the order of the matrix");
+  }
+
+  var wD = {0..<n, 0..<t};
+  var X = ones(wD);
+
+  if t > 1 {
+    for i in 1..<t {
+      X = column_resample(X, i);
+    }
+    for i in 0..<t {
+      while (column_needs_resampling(i, X)) {
+        X = column_resample(i, X);
+      }
+    }
+  }
+
+  X /= n;
+  //var usedInd = [0];
+  var oldEstimate = 0;
+  var S = zeros(wD);
+  var oldS = zeros(wD);
+  var k = 1;
+  //var ind = [];
+  //var bestColSofar: [0..<n] X.eltType;
+  var Y: [X.dom] X.eltType;
+  var ind_best = 0;
+
+  while true {
+    Y = dot(A, X);
+    // sum along axis zero taken on (n,t) dimension
+    // sparse matrix for a large n and a very small t
+    // this operation can be attributed to O(n) complexity.
+    var mags = sum(Y);
+    var (est, bestJ) = maxloc reduce zip(mags, mags.domain);
+    if est > oldEstimate || k == 2 {
+      if k>=2 {
+        ind_best = bestJ;
+      }
+      //var bestColSofar = Y[.., bestJ];
+    }
+    if k >= 2 && est <= oldEstimate {
+      est = oldEstimate;
+      break;
+    }
+    oldEstimate = est;
+    var oldS = S;
+    if k > itmax {
+      break;
+    }
+    S = sign_round_up(Y);
+
+    if is_X_Y_col_vectors_same(S, oldS) {
+      break;
+    }
+
+    if t>1 {
+      for i in 0..<t {
+        while column_needs_resampling(S, i, oldS){
+          column_resample(S,i);
+        }
+      }
+    }
+
+    var Z = dot(A.T, S);
+    var h = max(abs(Z), axis=1);
+
+    var maxHVal = max reduce(h);
+    if k>=2 && maxHVal == h[ind_best] {
+      break;
+    }
+
+  }
+  return oldEstimate;
+}
+
+proc argsort(A: [], param desc=false) {
+  use Sort;
+
+  record Cmp {
+
+  }
+
+  proc Cmp.compare(a, b) {
+    return a[0] - b[0];
+  }
+
+  var cmp: Cmp;
+  var B: [A.domain] int;
+  var n = B.size;
+  for i in 0..<n {
+    B[i] = i;
+  }
+
+  const AB = [ab in zip(A, B)] ab;
+  var i = 0;
+  for ab in AB.sorted(comparator=cmp) {
+    B[i] = ab[1];
+    i += 1;
+  }
+
+  if desc then
+    B.reverse();
+
+  return B;
+}
+
+proc is_X_Y_col_vectors_same(X: [?xD], Y: [?yD]){
+  var n = xD.dims(0);
+  for i in xD.dim(1) {
+    var b = X[.., i];
+    var fg = false;
+    for j in yD.dim(1) {
+      if dot(Y[.., j], b) == n {
+        fg = true;
+        break;
+      }
+    }
+    if !fg {
+      return false;
+    }
+  }
+  return true;
+}
+
+proc sign_round_up(A: [?D]){
+  forall (i,j) in D {
+    if b[i,j] == 0.0 then {
```

Comment:
is `b[i,j]` meant to be `A[i,j]`?

If so, there is no reason to loop over indices -- you can just loop over the array elements
```
forall a in A {
  <change a>
}
```
@lydia-duncan or @e-kayrakli might have even cleaner approaches.